GH-25118: [Python] Make NumPy an optional runtime dependency

raulcd opened the pull request:
### Rationale for this change

Being able to run pyarrow without requiring numpy.

### What changes are included in this PR?

If numpy is not present we are able to import pyarrow and run functionality.
A new CI job has been created to run some basic tests without numpy.

### Are these changes tested?

Yes via CI.

### Are there any user-facing changes?

Yes, NumPy can be removed from the user installation and pyarrow functionality still works

<!-- **This PR contains a "Critical Fix".** -->
* GitHub Issue: #25118

jorisvandenbossche commented on python/pyarrow/_compute.pyx:
```diff
@@ -473,7 +476,7 @@ cdef class MetaFunction(Function):
 cdef _pack_compute_args(object values, vector[CDatum]* out):
     for val in values:
-        if isinstance(val, (list, np.ndarray)):
+        if "numpy" in sys.modules and isinstance(val, (list, np.ndarray)):
```
For on the Python/cython side, I think we can have a simple globally used variable like `HAS_NUMPY` that is populated once (eg in lib.pyx), to avoid using the `"numpy" in sys.modules` pattern every time for checking if numpy is available

jorisvandenbossche commented on python/pyarrow/builder.pxi:
```diff
@@ -42,7 +42,7 @@ cdef class StringBuilder(_Weakrefable):
             The value to append to the string array builder.
         """
-        if value is None or value is np.nan:
+        if value is None or ('numpy' in sys.modules and value is np.nan):
```
Not directly related to this PR (and I assume this code is not used much), but it's also a bit strange that this checks explicitly for np.nan in addition to None (it could also check for `float("nan")`, for example), and additionally checking for NaN with `is np.nan` is also wrong because that doesn't always give true .. (eg `np.array([np.nan])[0] is np.nan` is False)

jorisvandenbossche commented on python/pyarrow/src/arrow/python/init.cc:
```diff
@@ -21,4 +21,15 @@
+  }
+}
+
+bool get_numpy_imported() { return numpy_imported; }
```
`is_numpy_imported`? (or `has_numpy` ?)

jorisvandenbossche commented on python/pyarrow/tests/test_adhoc_memory_leak.py:
```diff
@@ -28,6 +31,7 @@
     pass
 
 
+@pytest.mark.numpy
```
if something is already labeled as requiring pandas, it shouldn't need to be labeled as also requiring numpy?

jorisvandenbossche reviewed (commented):
I am a bit concerned about the maintenance effort for the tests and having to remember to add `@pytest.mark.numpy` everywhere needed. 
I am wondering: could we have a `nonumpy` marker instead, and label a subset of tests we explicitly want to run in the CI build without numpy? (of course, if that become a lot of tests, that will also be unwieldy) But then at least by default we don't have to think about marking it with numpy, and it's only when specifically working on tests for numpy being optional we have to add a marker.

Another idea would be to split certain test files in two where we could have a single marker at the top of the file indicating it's using numpy or not, instead of having to mark each individual test.

Just brainstorming a bit!

pitrou commented on python/pyarrow/tests/util.py:
```diff
@@ -22,7 +22,10 @@
 import decimal
 import gc
-import numpy as np
+try:
+    import numpy as np
+except ImportError:
+    pass
```
Is this import actually used?

raulcd replied:
I've done some cleanup here so we don't have to import here

pitrou commented on python/pyarrow/types.pxi:
```diff
@@ -149,14 +154,14 @@ def _is_primitive(Type type):
 def _get_pandas_type(arrow_type, coerce_to_ns=False):
     cdef Type type_id = arrow_type.id
-    if type_id not in _pandas_type_map:
+    if type_id not in _get_pandas_type_map():
```
Will this recreate the dict everytime this function is called? We don't want this.

amol- replied:
Might `functools.cache` be enough here? To avoid dealing with global data? Or are you concerned about the impact on first execution?

raulcd replied:
another possiblity if we want to do this at import time could be to add a new module with those definitions and only import them if numpy is present

pitrou commented on python/pyarrow/lib.pyx:
```diff
@@ -21,8 +21,10 @@
     import numpy as np
+    HAS_NUMPY=True
 except ImportError:
     np = None
```
What's the point of this? You can already write  `np is not None`. You don't have to define another internal variable.

jorisvandenbossche replied:
That was my suggestion to have a `HAS_NUMPY` or similar variable. This can then be used in other pyx files, I personally find that a bit easier to read than `np is not None` (but if in practice it turns out this is hardly used, might not be worth defining)

pitrou replied:
`HAS_NUMPY` is another internal variable that people have to remember, so I'm skeptical about it.

`<module name> is [not] None` is a widespread idiom for optional dependencies.

jorisvandenbossche replied:
To be clear I don't feel strongly about it, `np is not None` will also work just fine (although easy to understand as well, it's just not an idiom I am familiar with from other projects)

raulcd commented on python/pyarrow/tests/conftest.py:
```diff
@@ -111,11 +117,44 @@ def pytest_configure(config):
 
 
 def pytest_runtest_setup(item):
+    # Apply numpy mark if the test name already contains
```
@jorisvandenbossche what do you think about this? I am not sure we really want to do this automatically if the test name contains `numpy`, `pandas` or `ndarray` but it allows us to remove a bunch of marks from `test_array.py`:
See related commit: https://github.com/apache/arrow/pull/41904/commits/e378dd63c0eafc7983d54c8811e4da88a37f4b25

jorisvandenbossche replied:
Personally, I would say this feels a bit too magical .. It's nice to not have to add all those marks, but I think short term that is more explicit, and longer term I would maybe rather look into splitting test_array.py and putting all tests about conversion to/from numpy in a separate test file, so there are less tests remaining in test_array.py that require a mark.

raulcd replied:
I've decided to remove that and do the CI job only run some specific tests marked as "without_numpy". Those tests will run both with `numpy` and `nonumpy` but we can just select them with `-k without_numpy`.
Those will be just a small selection of the full test_suite.

raulcd commented on ci/scripts/python_test.sh:
```diff
@@ -69,4 +69,8 @@ export PYARROW_TEST_PARQUET_ENCRYPTION
 
 # Testing PyArrow
-pytest -r s ${PYTEST_ARGS} --pyargs pyarrow
+if [ -z "${TEST_COMMAND}" ]; then
```
`PYARROW_TEST_COMMAND` is not required anymore

raulcd commented:
@jorisvandenbossche @pitrou As part of this initial PR I've decided to create a new mark `pytest.mark.without_numpy` (feel free to propose a different name) that will be used to collect tests on the CI job when numpy is not present. The mark won't affect the tests when numpy is present and those tests will also be run.
I think we can follow up test refactoring or splitting some test modules (like `test_array.py`) in order to be able to run the full test suite with and without numpy in a follow up PR instead of trying to add it into this PR. I think this PR is already adding the functionality we are looking for and with the current tests being run and the new CI job we are testing basic functionality and validating compute, dataset, array are working without numpy.
I am marking this PR as ready for review now.

pitrou commented on ci/scripts/python_test.sh:
```diff
@@ -69,4 +69,4 @@ export PYARROW_TEST_PARQUET_ENCRYPTION
 
 # Testing PyArrow
-pytest -r s ${PYTEST_ARGS} --pyargs pyarrow
+pytest -r s ${PYTEST_ARGS} --pyargs ${PYTEST_PYARGS:-'pyarrow'}
```
This change doesn't look right, you have to follow `--pyargs` with something.

pitrou replied:
Ow, sorry, I misread `PYTEST_ARGS` vs. `PYTEST_PYARGS`

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -1566,6 +1568,9 @@ cdef class Array(_PandasConvertible):
         """
+        if "numpy" not in sys.modules:
+            raise ValueError(
+                "Cannot return a numpy.ndarray if Numpy is not present")
```
What should be our policy for errors here? `ValueError` could also be raised for other reasons. Perhaps `ImportError` would be more idiomatic? @jorisvandenbossche

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -1566,6 +1568,9 @@ cdef class Array(_PandasConvertible):
         -------
         array : numpy.ndarray
         """
+        if "numpy" not in sys.modules:
```
Why is this not using `np is not None`?
This is problematic because `sys.modules['numpy']` could actually be `None`, forcing NumPy to be considered absent (such as in unit tests).

pitrou commented on python/pyarrow/conftest.py:
```diff
@@ -20,7 +20,6 @@
 from pyarrow import Codec
 from pyarrow import fs
 
-import numpy as np
```
I think it would be nice to standardize on the `try/except` snippet that's used in other places.

pitrou commented on python/pyarrow/table.pxi:
```diff
@@ -495,6 +495,9 @@ cdef class ChunkedArray(_PandasConvertible):
         >>> n_legs.to_numpy()
         array([  2,   2,   4,   4,   5, 100])
         """
+        if "numpy" not in sys.modules:
```
Similar comments as above here (and below)...

pitrou commented on python/pyarrow/tests/test_flight.py:
```diff
@@ -54,6 +57,8 @@
 # Ignore these with pytest ... -m 'not flight'
 pytestmark = pytest.mark.flight
+# All tests on this module can be run when numpy is or is not available
+pytestmark = pytest.mark.without_numpy
```
But does this mean they wouldn't be run if NumPy is available? It is not clear to me as a reader of the code, and we risk ignoring tests by mistake.

pitrou commented on python/pyarrow/tests/test_without_numpy.py:
```diff
@@ -0,0 +1,58 @@
+
+# Marks all of the tests in this module
+# Ignore these with pytest ... -m 'not nonumpy'
+pytestmark = pytest.mark.nonumpy
```
So we have `pytest.mark.nonumpy` and `pytest.mark.without_numpy` and `pytest.mark.numpy`? How confusing is this?

jorisvandenbossche replied:
Maybe `does_not_require_numpy` instead of `without_numpy` ?

raulcd replied:
This has been changed to `numpy` and `nunumpy` as we have for pandas.

pitrou commented on docker-compose.yml:
```diff
@@ -1221,6 +1222,50 @@ services:
+        repo: ${REPO}
+        arch: ${ARCH}
+        python: ${PYTHON}
+        numpy: ${NUMPY}
+        pandas: ${PANDAS}
```
Well... ?

pitrou commented on docker-compose.yml:
```diff
@@ -1221,6 +1222,50 @@ services:
+      PARQUET_REQUIRE_ENCRYPTION:  # inherit
+      # The without numpy mark is only used to select tests to be run
+      # when numpy is not installed but those tests will also run
+      # if numpy is present.
```
But they would run anyway if they are not marked with `pytest.mark.numpy`, right? What is the point of this specific option?

pitrou replied:
I understand the concerns but I think the alternative solution that's been chosen (run a small subset of the tests) is much worse. Why wouldn't we exercise reading CSV and JSON files, for example?

I also find concerning that this is hidden somewhere in the `docker-compose.yml` (many developers will never go look there) instead of declared in plain sight at the top of the corresponding test files.

For example, if you don't want to run `test_csv.py` when NumPy is absent (assuming there's a good reason for it), you could add an explicit skip at the top of that file. The maintenance burden would be small enough.

jorisvandenbossche replied:
> Why wouldn't we exercise reading CSV

I know it was only an example, but FWIW it would be very easy to enable the CSV tests in this PR. The `test_csv.py` file does import numpy, but it only uses it once to generate random integers. That's something we should be able to replace with stdlib `random` quite easily I think.

pitrou replied:
Still, it would be better if the skips were done at the top of the test files themselves, instead of passing an environment variable from `docker-compose.yml`.

pitrou replied:
I don't know, if I add `test_csv` manually to `PYTEST_PYARGS`, things seem to work for me (though no CSV test gets run because of missing `without_numpy` markers):
```
+ pytest -r s -k without_numpy --pyargs pyarrow.tests.test_csv pyarrow.tests.test_compute pyarrow.tests.test_dataset pyarrow.tests.test_flight pyarrow.tests.test_without_numpy

[...]

==================================================================== test session starts ====================================================================
platform linux -- Python 3.8.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /
plugins: hypothesis-6.104.1, cython-0.2.2
collected 981 items / 882 deselected / 99 selected                                                                                                          

opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_compute.py ..............                                                         [ 14%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_dataset.py .........                                                              [ 23%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_flight.py .....s.............s.....s...........s.......ss..................ss.s.. [ 94%]
..                                                                                                                                                    [ 96%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_without_numpy.py sss                                                              [100%]

================================================================== short test summary info ==================================================================
SKIPPED [9] opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/conftest.py:94: slow NOT enabled
SKIPPED [3] opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/conftest.py:94: nonumpy NOT enabled
================================================ 87 passed, 12 skipped, 882 deselected, 13 warnings in 7.09s ================================================
```

pitrou replied:
And even if I clear `PYTEST_PYARGS`, the tests still pass:
```
+ pytest -r s -k without_numpy --pyargs pyarrow
/opt/conda/envs/arrow/lib/python3.8/site-packages/pytest_cython/__init__.py:2: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pkg_resources import get_distribution
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:3117: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:3117: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:2529: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(parent)
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:3117: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
==================================================================== test session starts ====================================================================
platform linux -- Python 3.8.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /
plugins: hypothesis-6.104.1, cython-0.2.2
collected 7634 items / 7527 deselected / 3 skipped / 107 selected                                                                                           

opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_array.py ........                                                                 [  7%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_compute.py ..............                                                         [ 20%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_dataset.py .........                                                              [ 28%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_flight.py .....s.............s.....s...........s.......ss..................ss.s.. [ 95%]
..                                                                                                                                                    [ 97%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_without_numpy.py sss                                                              [100%]

```

pitrou replied:
Ah, but I forgot to uninstall NumPy...

pitrou replied:
Ok, now I have:
```diff
diff --git a/python/pyarrow/tests/test_csv.py b/python/pyarrow/tests/test_csv.py
index 9ddb5197e9..3a77f1a2e2 100644
--- a/python/pyarrow/tests/test_csv.py
+++ b/python/pyarrow/tests/test_csv.py
@@ -36,7 +36,12 @@ import weakref
 
 import pytest
 
-import numpy as np
+pytestmark = pytest.mark.numpy
+
+try:
+    import numpy as np
+except ImportError:
+    np = None
 
 import pyarrow as pa
 from pyarrow.csv import (
```

And then I can collect `test_csv` without NumPy:
```
+ pytest -r s -k without_numpy --pyargs pyarrow.tests.test_array pyarrow.tests.test_csv

[...]

==================================================================== test session starts ====================================================================
platform linux -- Python 3.8.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /
plugins: hypothesis-6.104.1, cython-0.2.2
collected 420 items / 412 deselected / 8 selected                                                                                                           

opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_array.py ........                                                                 [100%]

````

pitrou replied:
By the way, the fact that only those tests listed in `PYTEST_PYARGS` *and* with a `pytest.mark.without_numpy` marker are executed in this instance is clearly neither maintainable nor future-proof (we'll easily forget about adding these when adding new tests).

jorisvandenbossche replied:
> And then I can collect test_csv without NumPy:

`test_csv.py` is an easy example because it hardly uses numpy in practice (as I mentioned above, there is exactly one use of it to generate random integers, that can be easily be fixed in this PR). But the problem is for files that use numpy in `pytest.mark.parametrize`, as far as I understand

raulcd replied:
yeah, `test_convert_builtin.py` is a challenge for example

pitrou replied:
I've just tested this incantation and it seems to workaround the test collection problem (unlike `pytestmark = ...` which would still execute the module to the end):
```python
try:
    import numpy as np
except ImportError:
    pytest.skip(reason="blah", allow_module_level=True)
```

jorisvandenbossche replied:
> I've just tested this incantation and it seems to workaround the test collection problem

Ah, that would be a nice solution!

raulcd replied:
this is indeed really nice! Thanks @pitrou I'll work with that! Will ping you once I think this is again ready for review!
Thanks both for the constructive feedback and help!

pitrou replied:
Thanks for your work @raulcd :-)

pitrou commented on docker-compose.yml:
```diff
@@ -1221,6 +1222,50 @@ services:
+      # when numpy is not installed but those tests will also run
+      # if numpy is present.
+      PYTEST_ARGS: "-k without_numpy"
+      # Some tests fail to be collected due to numpy fixtures.
```
Can we create a followup GH issue for this and reference it here?

pitrou commented on docker-compose.yml:
```diff
@@ -1221,6 +1222,50 @@ services:
+      # if numpy is present.
+      PYTEST_ARGS: "-k without_numpy"
+      # Some tests fail to be collected due to numpy fixtures.
+      # That is why we just collect tests from the following modules.
```
That's a disturbingly small set of test modules.

pitrou commented:
I'm really -1 on the current testing approach where everything is disabled by default and tests have to be whitelisted **twice** to be executed.

jorisvandenbossche commented:
> tests have to be whitelisted **twice**

We could also only do the file selection, and for those files remove the marker (which means that either all tests in their have to run without numpy, or be marked with `pytest.mark.numpy`)

And then we can in follow-ups gradually make that list of files larger until it's all test files and we can simply run pytest without a specific file selection.

raulcd commented:
> I'm really -1 on the current testing approach where everything is disabled by default and tests have to be whitelisted **twice** to be executed.

Thanks @pitrou as pointed on the PR this was just an initial (possible) solution to have some tests being run, I am ok if we want to remove the `without_numpy` mark but I would still prefer to only run some specific modules.

raulcd commented:
I still have to do a couple of things before is ready for review again but after going over all tests we have 190 tests marked as requiring numpy:
```
~/code/arrow/python  (GH-25118)  $ grep -R "@pytest.mark.numpy" . | wc -l
190
```
Even though it seems a lot we have 268 for pandas:
```
~/code/arrow/python  (GH-25118)  $ grep -R "@pytest.mark.pandas" . | wc -l
268
```
With those changes we are running the following tests without numpy:
```
2258 passed, 934 skipped, 17 xfailed, 1 xpassed in 136.27s
```

raulcd commented:
@pitrou @jorisvandenbossche I've gone through all the review comments and acted upon them, rebased and fixed some minor related CI issues.
Both the jobs with and without numpy scans and run all relevant tests. We only have two marks: `numpy` and `nonumpy` which work the same as it does with the `pandas` and `nopandas` today.
This is ready for a new round of review.
Thanks for your help!

amol- commented on python/pyarrow/_compute.pyx:
```diff
@@ -473,7 +476,7 @@ cdef class MetaFunction(Function):
 cdef _pack_compute_args(object values, vector[CDatum]* out):
     for val in values:
-        if isinstance(val, (list, np.ndarray)):
+        if isinstance(val, list) or (np is not None and isinstance(val, np.ndarray)):
```
I suggest we make a `SUPPORTED_INPUT_LIST_TYPES` (happy with any better name if available) and change the `isinstance` checks to be `isinstance(val, SUPPORTED_INPUT_LIST_TYPES)` 

then use
```
SUPPORTED_INPUT_LIST_TYPES = (list, tuple)
try:
  import numpy as np
except ImportError:
  np = None
else:
  SUPPORTED_INPUT_LIST_TYPES += (np.ndarray, )
```

That would avoid the isinstance check in every single place and would make it more reasable the `if`s

amol- commented on python/pyarrow/lib.pyx:
```diff
@@ -32,8 +35,12 @@ from pyarrow.includes.common cimport PyObject_to_object
-arrow_init_numpy()
+# Initialize NumPy C API only if numpy was able to be imported
+
+
```
stray newlines?

amol- commented on python/pyarrow/pandas_compat.py:
```diff
@@ -85,27 +89,32 @@ def get_logical_type(arrow_type):
+    numpy_logical_type_map = get_numpy_logical_type_map()
     try:
-        return _numpy_logical_type_map[pandas_collection.dtype.type]
+        return numpy_logical_type_map[pandas_collection.dtype.type]
```
I suggest we make the `get_numpy_logical_type_map` throw a proper error when `numpy` is not available. Currently this check would throw a `KeyError` and would do pretty random things with anything that has a `dtype` attribute. 

Same for `get_pandas_logical_type_map`

jorisvandenbossche commented:
@github-actions crossbow submit -g python

jorisvandenbossche commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -4594,6 +4604,7 @@ def test_write_table_multiple_fragments(tempdir):
     )
 
 
+@pytest.mark.numpy
 def test_write_iterable(tempdir):
     table = pa.table([
         pa.array(range(20)), pa.array(np.random.randn(20)),
```
Some of those `np.random.randn(20)` calls could be replaced with an utility using the stdlib? (can also be for a follow-up)

jorisvandenbossche commented on python/pyarrow/tests/test_io.py:
```diff
@@ -1106,11 +1119,13 @@ def _check_native_file_reader(FACTORY, sample_data,
     assert f.tell() == ex_length
 
 
+@pytest.mark.numpy
 def test_memory_map_reader(sample_disk_data):
```
Another to do item for a follow-up to rewrite `sample_disk_data` to not use numpy

jorisvandenbossche commented on python/pyarrow/tests/test_ipc.py:
```diff
@@ -161,14 +164,17 @@ def test_empty_file():
         pa.ipc.open_file(pa.BufferReader(buf))
 
 
+@pytest.mark.numpy
 def test_file_simple_roundtrip(file_fixture):
```
Another to do: update `IpcFixture.write_batches` to not use numpy (that would avoid a lot of marks in this file, I think)

jorisvandenbossche commented on python/pyarrow/tests/test_table.py:
```diff
@@ -972,65 +977,67 @@ def check_tensors(tensor, expected_tensor, type, size):
     assert tensor.strides == expected_tensor.strides
 
 
-@pytest.mark.parametrize('typ', [
-    np.uint8, np.uint16, np.uint32, np.uint64,
-    np.int8, np.int16, np.int32, np.int64,
-    np.float32, np.float64,
-])
```
If we would want to keep this parametrization, could easily use strings instead (and then inside the function convert that with `np.dtype(..)` or so). 
But no strong opinion, iterating over the options inside the test is also fine.

jorisvandenbossche reviewed (commented):
Looking good! Did a detailed review now

raulcd commented on python/pyarrow/tests/test_cpp_internals.py:
```diff
@@ -26,6 +31,10 @@ def inject_cpp_tests(ns):
     for case in get_cpp_tests():
+        if np is None and ('numpy' in case.name or 'pandas' in case.name):
+            # Skip C++ tests that require pandas or numpy if numpy not present
+            continue
```
As those are not test items at that point I wasn't sure how to add the markers here. I ended up adding a `pytest_collection_modifyitems` on `conftest.py` to add the corresponding markers. Let me know if this approach makes sense to you.

jorisvandenbossche replied:
I am not entirely sure this will work, but since you are essentially defining the test function here, I was wondering if something like this would work:

```
        def wrapper(request, case=case):
            if 'numpy' in case.name:
                request.node.add_marker(pytest.mark.numpy)
            case()
```

(and similar for pandas)

jorisvandenbossche replied:
It might be this does not work for _selecting_ tests, though, because the marker is only added when running the test (it does work for xfailing or skipping). 

So could either use that to skip the test directly (instead of through the `numpy` mark), or could also do something like the following (but that is more verbose):

```
if 'numpy' in case.name:

    @pytest.mark.numpy
    def wrapper():
        case()

else:

    def wrapper():
        case()
```

In the end this is also a decorator, so maybe the following also works (not entirely sure this works this way as those pytest decorators are a bit magical ..):

```
def wrapper():
    case()

if 'numpy' in case.name:
    wrapper = pytest.mark.numpy(wrapper)
```

raulcd replied:
Thanks @jorisvandenbossche ! I've tested the proposed approach and validated it works. I've gone with that removing the `pytest_collection_modifyitems`

jorisvandenbossche replied:
Nice!

raulcd commented on python/pyarrow/tests/strategies.py:
```diff
@@ -276,7 +279,7 @@ def arrays(draw, type, size=None, nullable=True):
         # Workaround ARROW-4952: no easy way to assert array equality
         # in a NaN-tolerant way.
-        values[np.isnan(values)] = -42.0
+        values[math.isnan(values)] = -42.0
```
it seems the hypothesis test failed, I've reverted back to `np.isnan` for those.

raulcd commented on python/pyarrow/tests/test_csv.py:
```diff
@@ -415,6 +418,7 @@ def test_header_skip_rows(self):
         }
 
+    @pytest.mark.numpy
     def test_skip_rows_after_names(self):
```
I decided to refactor `make_random_csv` to not require numpy

jorisvandenbossche commented:
@github-actions crossbow submit test-conda-python-3.11-hypothesis

raulcd commented on python/pyarrow/tests/test_array.py:
I've created a follow up issue here: https://github.com/apache/arrow/issues/43510

raulcd commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -23,8 +23,12 @@
+    import numpy as np
+except ImportError:
+    pytest.skip(reason="Failures on test collection due to numpy NOT enabled",
+                allow_module_level=True)
```
I've ended up adding the following commit: https://github.com/apache/arrow/pull/41904/commits/ef97ad314ba7b5990c0d580f923f715908e9db1a
to remove the file skip and to individual markers per test. If you could take a look on that one.

raulcd commented on python/pyarrow/tests/parquet/test_data_types.py:
```diff
@@ -331,6 +335,7 @@ def test_column_of_lists(tempdir):
     tm.assert_frame_equal(df, df_read)
 
 
+@pytest.mark.numpy
```
I think we want to do it. I've already donde it for a bunch of tests.

raulcd commented:
@github-actions crossbow submit -g python

raulcd commented:
CI test failures are unrelated

jorisvandenbossche commented on python/pyarrow/tests/conftest.py:
```diff
@@ -25,8 +25,14 @@
 import pytest
 import hypothesis as h
+try:
+    import numpy as np
```
Is this import still needed? (I think with the latest updates `np` is not used anymore in this file?)

jorisvandenbossche commented on python/pyarrow/tests/strategies.py:
```diff
@@ -35,7 +38,10 @@
     except ImportError:
         zoneinfo = None
-import numpy as np
+try:
+    import numpy as np
+except ImportError:
+    np = None
```
The "strategies" defined in this file are not marked as needing numpy. Are the tests using those strategies marked as such? 

(not that we are testing this, because we run the hypothesis tests only in one build which has numpy, so it's also fine to assume that numpy is available for the hypothesis tests I think)

raulcd replied:
I've run hypothesis tests locally and have marked the ones that require numpy explicitly

jorisvandenbossche commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -285,15 +288,16 @@ def test_list_with_non_list(seq):
 @pytest.mark.parametrize("factory", [
     pa.list_, pa.large_list, pa.list_view, pa.large_list_view])
 def test_nested_arrays(seq, factory):
-    arr = pa.array(seq([np.array([], dtype=np.int64),
-                        np.array([1, 2], dtype=np.int64), None]),
+    arr = pa.array(seq([pa.array([], type=pa.int64()),
+                        pa.array([1, 2], type=pa.int64()), None]),
```
In hindsight, this should maybe be kept as numpy arrays .. (and just mark as requiring numpy)  
At least it is useful to also test nested data construction where the inner values are numpy arrays (probably ideally we would test both, like parametrize on the sequence type for the inner sequences as well?)

raulcd replied:
I've parametrized the test with inner_sequence reusing the `sequence_types` defined

jorisvandenbossche reviewed (approved):
Looks great!

raulcd commented:
@github-actions crossbow submit test-conda-python-3.11-hypothesis

pitrou commented on python/pyarrow/_compute.pyx:
```diff
@@ -473,7 +481,7 @@ cdef class MetaFunction(Function):
     for val in values:
-        if isinstance(val, (list, np.ndarray)):
+        if isinstance(val, SUPPORTED_INPUT_ARR_TYPES):
             val = lib.asarray(val)
```
Unrelated to this PR, but at some point we should accept any [ArrowArrayExportable](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html#arrowarray-export)? @jorisvandenbossche

jorisvandenbossche replied:
We should do that indeed! 

(https://github.com/apache/arrow/issues/43410 is a general issue about accepting such objects in more places)

pitrou commented on python/pyarrow/tests/interchange/test_conversion.py:
```diff
@@ -16,11 +16,16 @@
+    import numpy as np
+except ImportError:
+    pytest.skip(reason="Failures on test collection due to numpy NOT enabled",
+                allow_module_level=True)
```
Perhaps open a GH issue to fix this, and reference it here?

jorisvandenbossche replied:
Looking at the file here, it seems the actual parametrization using numpy are just float32/float64 dtypes, and those could easily be converted to strings

pitrou commented on python/pyarrow/tests/interchange/test_interchange_spec.py:
```diff
@@ -19,10 +19,14 @@
+    import numpy as np
+except ImportError:
+    pytest.skip(reason="Failures on test collection due to numpy NOT enabled",
+                allow_module_level=True)
```
Ditto here and in other similar occurrences.

pitrou commented on python/pyarrow/tests/test_csv.py:
```diff
@@ -54,18 +53,31 @@ def generate_col_names():
+
 def make_random_csv(num_cols=2, num_rows=10, linesep='\r\n', write_names=True):
-    arr = np.random.RandomState(42).randint(0, 1000, size=(num_cols, num_rows))
+    arr = [random.randint(0, 1000) for _ in range(num_cols * num_rows)]
```
We want to make random generation deterministic just like above, so perhaps something like:
```suggestion
    rnd = random.Random(42)
    arr = [rnd.randint(0, 1000) for _ in range(num_cols * num_rows)]
```

pitrou commented on python/pyarrow/tests/test_cython.py:
```diff
@@ -80,6 +80,7 @@ def check_cython_example_module(mod):
         mod.cast_scalar(scal, pa.list_(pa.int64()))
 
 
+@pytest.mark.numpy
```
Hmm, is there anything that requires NumPy in this test?

raulcd replied:
there is an import numpy on the cython `setup_template`, it fails with:
```
pyarrow/tests/test_cython.py::test_cython_api Traceback (most recent call last):
  File "/tmp/pytest-of-raulcd/pytest-243/test_cython_api0/setup.py", line 5, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
FAILED
```
If I try to remove `numpy` (which I don't think we want on this test) the `pyarrow_cython_example` file also fails to compile with:

```
x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/home/raulcd/code/arrow/python/pyarrow/include -I/home/raulcd/code/no-numpy/include -I/usr/include/python3.10 -c pyarrow_cython_example.cpp -o build/temp.linux-x86_64-3.10/pyarrow_cython_example.o -std=c++17
pyarrow_cython_example.cpp:1380:10: fatal error: numpy/halffloat.h: No such file or directory
 1380 | #include "numpy/halffloat.h"
      |          ^~~~~~~~~~~~~~~~~~~
```

pitrou replied:
I see, so NumPy is still required for Cython integration until we make NumPy optional in our headers too (but we might not care much about this).

pitrou replied:
Can you just add a comment about it?

jorisvandenbossche replied:
My understanding is that this PR makes numpy optional at _runtime_, but numpy is indeed still a build-time dependency. But indeed, that doesn't seem to be a big priority as at build time numpy should be a relatively easy dependency to install.

raulcd replied:
I've added a comment

pitrou commented on python/pyarrow/tests/test_dataset_encryption.py:
```diff
@@ -170,6 +173,7 @@ def test_write_dataset_parquet_without_encryption():
         _ = pformat.make_write_options(encryption_config="some value")
 
 
+@pytest.mark.numpy
```
Hmm, it should be easy to remove NumPy usage from this test, no?

pitrou commented on python/pyarrow/tests/test_types.py:
```diff
@@ -1261,19 +1264,22 @@ def test_field_modified_copies():
     assert f0.equals(f0_)
 
 
+@pytest.mark.numpy
 def test_is_integer_value():
     assert pa.types.is_integer_value(1)
     assert pa.types.is_integer_value(np.int64(1))
     assert not pa.types.is_integer_value('1')
```
Perhaps we should instead just guard the corresponding assertion?
```suggestion
@pytest.mark.numpy
def test_is_integer_value():
    assert pa.types.is_integer_value(1)
    if np is not None:
        assert pa.types.is_integer_value(np.int64(1))
    assert not pa.types.is_integer_value('1')
```

pitrou commented on python/pyarrow/tests/util.py:
```diff
@@ -110,27 +109,16 @@ def randdecimal(precision, scale):
 
 
 def random_ascii(length):
+    import numpy as np
     return bytes(np.random.randint(65, 123, size=length, dtype='i1'))
```
Perhaps this can be fixed to not require NumPy?
```suggestion
    return bytes([random.randint(65, 122) for i in range(length)])
```

raulcd commented:
@pitrou @jorisvandenbossche I think I've gone through all the comments on the PR. I will be off for some days since Thursday, let me know if there's anything else you want me to change before that :)

pitrou commented on python/pyarrow/tests/test_csv.py:
```diff
@@ -54,18 +53,32 @@ def generate_col_names():
 
+def split_rows(arr, num_cols, num_rows):
+    # Split a num_cols x num_rows array into rows
+    for i in range(0, num_rows*num_cols, num_cols):
+        yield list(itertools.islice(arr, i, i + num_cols))
```
I've just noticed that you're going through `itertools.islice` just to accumulate the results into a list, so you can probably slice the input directly, which is going to be much more efficient:
```suggestion
    for i in range(0, num_rows * num_cols, num_cols):
        yield arr[i:i + num_cols]
```
```

pitrou commented on python/pyarrow/tests/test_csv.py:
```diff
@@ -54,18 +53,32 @@ def generate_col_names():
+
+def split_columns(arr, num_cols, num_rows):
+    # Split a num_cols x num_rows array into columns
+    for i in range(0, num_cols):
+        yield list(itertools.islice(arr, i, num_cols * num_rows, num_cols))
```
Same suggestion here:
```suggestion
    for i in range(0, num_cols):
        yield arr[i::num_cols]
```

pitrou commented on python/pyarrow/tests/test_types.py:
```diff
@@ -1263,17 +1266,20 @@ def test_field_modified_copies():
     assert not pa.types.is_integer_value('1')
 
 
+@pytest.mark.numpy
```
Can we just apply the same principle here and below (guard NumPy-specific checks with `if np is not None` instead of skipping the whole test if NumPy is absent)?

jorisvandenbossche commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -102,6 +100,37 @@ def test_basics(value, ty, klass, pickle_module):
 
+@pytest.mark.numpy
+def test_basics_np_required(pickle_module):
+    value, ty, klass = np.float16(1.0), pa.float16(), pa.HalfFloatScalar
+    s = pa.scalar(value, type=ty)
```
What we are testing here essentially is that a numpy float scalar (`value`) can be converted to a pyarrow Scalar with `pa.scalar(..)`. But all the rest of the test is not specific to that fact that we created it from a numpy scalar, so I don't think you need to repeat that entirely here (the test above already covers all those things for HalfFloatScalar), so I think you can simplify the test here.

jorisvandenbossche replied:
And also, it seems this conversion from numpy scalar to pyarrow scalar is also covered in the test below (`test_numerics`) ?

raulcd replied:
I am not sure I understand how the test above is testing `HalfFloatScalar` if I remove from the parametrization:
`(np.float16(1.0), pa.float16(), pa.HalfFloatScalar),`
I am trying to remove `np` from parametrize in order to avoid collection failures.
@jorisvandenbossche are you suggesting I can just remove `(np.float16(1.0), pa.float16(), pa.HalfFloatScalar),` as this one is already covered on the rest of the tests and avoid adding a new test for this case?

jorisvandenbossche replied:
> I am not sure I understand how the test above is testing `HalfFloatScalar` if I remove from the parametrization:
> `(np.float16(1.0), pa.float16(), pa.HalfFloatScalar),`

Ah, sorry, I assumed that it would still be in the paramatrization, as we are also testing the others without a numpy scalar. Is there a reason the test above cannot use `(1.0, pa.float16(), pa.HalfFloatScalar)` for the parametrization?

jorisvandenbossche replied:
And to answer my own question: because we cannot create a half float scalar right now from a plain python float:

```
In [7]: pa.scalar(1.0, pa.float16())
...
ArrowTypeError: Expected np.float16 instance
```

OK, you can ignore my comment here ;)

jorisvandenbossche replied:
Can you maybe just add a comment at the beginning of `test_basics_np_required` explaining the context (it's duplicating `test_basics` but for float16, which requires a numpy scalar to create it)

raulcd replied:
Thanks @jorisvandenbossche I've added a comment

raulcd commented:
CI failures are unrelated. @jorisvandenbossche @pitrou I've gone through the last comments again. Sorry, I had to rebase in order to fix some issues. The last commits, from https://github.com/apache/arrow/pull/41904/commits/a566aeca944707f31f0ad93b06651948aca88174 onwards, are the new commits.

jorisvandenbossche commented:
All good from my side!

While this PR does make it possible to use pyarrow at runtime without numpy installed, it does not actually makes the _dependency_ optional, in a standard packaging sense. Do we want to discuss that in a follow-up issue whether we also want to actually remove numpy from the required dependencies list?

raulcd commented:
> Do we want to discuss that in a follow-up issue whether we also want to actually remove numpy from the required dependencies list?

I was planning to create a follow up issue where we remove numpy from the required dependencies list. I've just created: https://github.com/apache/arrow/issues/43846

raulcd commented:
I've rebased again to keep it up to date with main. @pitrou do you have any other comments / concerns before merging?

pitrou commented:
@github-actions crossbow submit -g wheel -g python

pitrou reviewed (approved):
LGTM, let's just wait for CI. Great work Raul!

jorisvandenbossche commented:
Thanks @raulcd!
